pkg/ride/functions_invocations.go

```diff
@@ -0,0 +1,33 @@
+package ride
+
+import (
+	"github.com/wavesplatform/gowaves/pkg/proto"
+)
+
+func invokeFunctionFromDApp(env environment, recipient proto.Recipient, fnName rideString, listArgs rideList) (Result, error) {
+	newScript, err := env.state().GetByteTree(recipient)
+	if err != nil {
+		return nil, EvaluationFailure.Wrap(err, "failed to get script by recipient")
+	}
+	tree, err := Parse(newScript)
+	if err != nil {
+		return nil, EvaluationFailure.Wrap(err, "failed to parse script")
+	}
+	if tree.LibVersion < 5 {
+		return nil, RuntimeError.Errorf("failed to call 'invoke' for script with version %d. Scripts with version 5 are only allowed to be used in 'invoke'", tree.LibVersion)
+	}
+	args, err := convertListArguments(listArgs)
+	if err != nil {
+		return nil, EvaluationFailure.Wrapf(err, "failed to invoke function '%s'", fnName)
+	}
+	e, err := treeFunctionEvaluator(env, tree, string(fnName), args)
+	if err != nil {
+		return nil, EvaluationFailure.Wrapf(err, "failed to call function '%s'", fnName)
+	}
+	res, err := e.evaluate()
+	if err != nil {
+		// Evaluation failed we have to return a DAppResult that contains spent execution complexity
+		return DAppResult{complexity: e.complexity}, err
```

Comment:
Create third function return parameter for execution complexity. Result must be ignored in case of error.

Reply:
What's the difference? If I have to ignore one result, what holds me from ignoring second?

Reply:
Hmm, sounds reasonable